### PR TITLE
RHIROS-597 | Executive report PDF - DB Schema and Processor updates

### PR DESCRIPTION
**Pull request by saltgen**
### Changes:

1. Model updates
2. Handler method for populating db records
3. Added a constants.py file
4. unittest

### CODE REVIEW

**Comment by saltgen** [comment]
Need to add changes for `idle` state, hence the `Do Not Merge` label

**Comment by kgaikwad** [comment]
IIUC, this DB schema change should be based on Advisor side sub_states change - https://gitlab.cee.redhat.com/insights-rules/insights-plugins/-/merge_requests/3951.
@saltgen  - could you check latest comments and confirm any further DB schema changes required?

**Comment by saltgen** [comment]
> IIUC, this DB schema change should be based on Advisor side sub_states change - https://gitlab.cee.redhat.com/insights-rules/insights-plugins/-/merge_requests/3951. @saltgen - could you check latest comments from MR-3951 and confirm any further DB schema changes required?

@kgaikwad Schema changes pushed considering such sample response, [here](https://gitlab.cee.redhat.com/insights-rules/insights-plugins/-/merge_requests/3951#note_3860086)

**Comment by r14chandra** [comment]
@saltgen I don't see any constants.py file here. Although this file has been imported in this PR #208

**Comment by saltgen** [comment]
> @saltgen I don't see any constants.py file here. Although this file has been imported in this PR #208

@r14chandra Good catch! The file got ignored by my git for some reason, I have pushed it on #208

File: migrations/versions/e65ffbc90def_.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# revision identifiers, used by Alembic.
+revision = 'e65ffbc90def'
+down_revision = '77fdc857c94f'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('systems', sa.Column('cpu_state', sa.ARRAY(sa.String()), nullable=True))
```
[the remark this answers is not in the document]
Comment (saltgen) [reply migrations/versions/e65ffbc90def_.py:21]: @kgaikwad Do we need substates data in performance profile?
FYI I have added changes to remove substates data from getting saved into `rule_hit_details`, [here](https://github.com/RedHatInsights/ros-backend/pull/203/files#diff-0c17bcf0cdabfe68b39e608007e0195be1c3ae63f6395c4f9bd809e003089e26R131)

File: tests/test_insights_engine_result_consumer.py [hunk 2]
```diff
@@ -172,3 +172,26 @@ def test_history_record_creation(engine_result_message, engine_consumer, db_setu
             system_id=system_record.id
         ).count()
         assert history_rec_count == 1
+
+
+def test_substate_updates(engine_result_message, engine_consumer, db_setup, performance_record):
+    engine_result_message = engine_result_message("insights-engine-result-under-pressure.json")
```
[the remark this answers is not in the document]
Comment (saltgen) [reply tests/test_insights_engine_result_consumer.py:178]: @r14chandra Felt unnecessary, it was required for the test in API, hence added the same [here](https://github.com/RedHatInsights/ros-backend/pull/208/files#diff-ef7e906a1ad4cd29b9a2ccc2dc5355068a0f17eca3b8913842cfa0725ecf0209R58-R60)

[the remark this answers is not in the document]
Comment (kgaikwad) [reply tests/test_insights_engine_result_consumer.py:178]: I would also prefer to update sample/test data files content that matches current rules implementation.

File: migrations/versions/e65ffbc90def_.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (saltgen) [reply migrations/versions/e65ffbc90def_.py:21]: @kgaikwad That is correct, so `substates` should be present in `reports`, I have removed the `.pop()`,
such that pp and pph can both have `states`(i.e. substates) inside `rule_hit_details`.
I hope this is okay, change pushed.

File: tests/test_insights_engine_result_consumer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (saltgen) [reply tests/test_insights_engine_result_consumer.py:178]: @kgaikwad  @r14chandra Added sample-files/sample-substates.json :+1: 
Change pushed

[the remark this answers is not in the document]
Comment (r14chandra) [reply tests/test_insights_engine_result_consumer.py:178]: I don't see this file here

[the remark this answers is not in the document]
Comment (saltgen) [reply tests/test_insights_engine_result_consumer.py:178]: @r14chandra Re-pushed, could you check now?

File: sample-files/sample-substates.json [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                    ],
+                    [
+                        "g4dn.16xlarge",
+                        5.875
+                    ]
+                ],
+                "states": {
+                    "cpu": [
+                        "CPU_UNDERSIZED",
+                        "CPU_UNDERSIZED_BY_PRESSURE"
```
Comment (apuntamb) [inline sample-files/sample-substates.json:67]: Are these two considered individual sub-states ?

Comment (saltgen) [reply sample-files/sample-substates.json:67]: @apuntamb Yes, one is `Undersized` and another `Under Pressure`

**Comment by saltgen** [comment]
Found a bug, substates are not getting refreshed on system refreshes, hence Do Not Merge label

File: migrations/versions/e65ffbc90def_.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    op.add_column('systems', sa.Column('io_state', sa.ARRAY(sa.String()), nullable=True))
+    op.add_column('systems', sa.Column('memory_state', sa.ARRAY(sa.String()), nullable=True))
```
Comment (kgaikwad) [inline migrations/versions/e65ffbc90def_.py:23]:
```suggestion
    op.add_column('systems', sa.Column('cpu_states', sa.ARRAY(sa.String()), nullable=True))
    op.add_column('systems', sa.Column('io_states', sa.ARRAY(sa.String()), nullable=True))
    op.add_column('systems', sa.Column('memory_states', sa.ARRAY(sa.String()), nullable=True))
```
These fields are of Array type and can include multiple values so does it make more sense to keep names as `cpu_states/memory_states/io_states`

**Review by kgaikwad — CHANGES_REQUESTED** [review]

File: migrations/versions/e65ffbc90def_.py (hunk 4 above)
Comment (saltgen) [reply migrations/versions/e65ffbc90def_.py]: @kgaikwad Thanks! Changes :rocket:

**Review by kgaikwad — CHANGES_REQUESTED** [review]

**Comment by kgaikwad** [comment]
Merging this PR, unresolved comments we will track in separate PR.

**Comment by kgaikwad** [comment]
Merging this PR, unresolved comments we will track in separate PR.

**Pull request merged**